HPCC-20332 Initial Dali XSD

kenrowland opened the pull request:
Convert Dali to new format
Change XSDs to use new default custom namespace attributes
Add hpcc:hiddenIf custom namespace attribute

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>


<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [ ] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

kenrowland commented:
@rpastrana Please review. 
Note this PR contains the change for the new default specification. All XSDs had to be changed along with the use in Dali

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: eff4950b918f2074affe58a4a971b86169706032
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   15 sec (00:00:15) |  172 sec (00:02:52) |   93 sec (00:01:33) |   21 sec (00:00:21) |   43 sec (00:00:43) | 1332 sec (00:22:12) |   19 sec (00:00:19) | 1695 sec (00:28:15) |

rpastrana commented:
@kenrowland this needs to be rebased

kenrowland commented:
@rpastrana This PR is based on previous PRs and one active PR. I would only be able to eliminate 1 commit at this time because the other PR is not merged. I would prefer to keep the changes isolated so I can manage them better. The PR will of course be fully rebased prior to a merge request. Is it OK if you only look at the last commit? It represents all the changes for the subject of the PR

rpastrana commented on configuration/config2/XSDSchemaParser.cpp:
```diff
@@ -719,22 +719,63 @@ void XSDSchemaParser::parseAllowedValue(const pt::ptree &allowedValueTree, Schem
+    }
+    else
+    {
+        pCfgValue->setHiddenIf(hiddenIf + visibleIf);
```
what is actually being done here? 
we know one of hiddenIf or visibleIf is empty, but why are you appending hiddenif and visibleIf ?

rpastrana replied:
this is going to cause maintainability problems, let's rewrite using verbose if else cluases

kenrowland replied:
Changed to use a ternary operator so it's clear which path attribute is being used.

rpastrana commented on configuration/config2/XSDSchemaParser.cpp:
```diff
@@ -719,22 +719,63 @@ void XSDSchemaParser::parseAllowedValue(const pt::ptree &allowedValueTree, Schem
+    }
+    else if (!preset.empty())
+    {
+        pCfgValue->setCodeDefault(preset);
```
why is this function called setCodeDefault, it should be setPreset

kenrowland replied:
The presetValue and forcedConfigValue XSD attributes are for the XSDs so that they attributes are clear as to what they mean. The ESP interface still uses a default in code naming convention, therefore I kept the naming convention throughout the code so that the above was limited to the XSD and parsing. All other code remained unchanged.

rpastrana replied:
I see, let's get all interfaces in synch w/ the new naming convention

kenrowland replied:
Agreed. Will change in a later PR to ensure consistency of naming convention

rpastrana commented on configuration/config2/XSDSchemaParser.cpp:
```diff
@@ -719,22 +719,63 @@ void XSDSchemaParser::parseAllowedValue(const pt::ptree &allowedValueTree, Schem
+    }
+    else
+    {
+        pCfgValue->setDefaultValue(forcedConfigValue);
```
correct the method name

kenrowland replied:
See previous comment.

rpastrana commented on initfiles/componentfiles/configschema/xsd/dafilesrv.xsd:
```diff
@@ -1,13 +1,13 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <xs:schema
-    xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormDefault="qualified" attributeFormDefault="unqualified"
+    xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormhpcc:presetValue="qualified" attributeFormhpcc:presetValue="unqualified"
```
make sure the component owner reviews the xsd, I am not reviewing the xsd

kenrowland replied:
that is the next step. I will open a JIRA for that specifically.

rpastrana reviewed (changes requested):
@kenrowland  make sure the XSDs are reviewed by component experts

kenrowland reviewed (commented):
@rpastrana Please see comments

kenrowland commented:
@rpastrana Just a note that the exp.xsd conflict will be resolved once PR 11554 is merged. The PR is dependent on that PR.

richardkchapman commented:
Please rebase

kenrowland commented:
@richardkchapman I am waiting for PR 11554 to be merged first as there are dependencies on that PR.

kenrowland commented:
@rpastrana If there are no more comments, please approve

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: abcc6d5c08305836f7a3ba919cb91e20c113f703
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |806 |806 |0 |
|test (thor) |730 |730 |0 |
|test (roxie) |879 |879 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   11 sec (00:00:11) |  189 sec (00:03:09) |   94 sec (00:01:34) |   20 sec (00:00:20) |   42 sec (00:00:42) | 1350 sec (00:22:30) |   18 sec (00:00:18) | 1724 sec (00:28:44) |

rpastrana reviewed (changes requested):
@kenrowland a couple of comments

kenrowland reviewed (commented):
@rpastrana Updated per request. See comments and code change.

rpastrana reviewed (approved):
@kenrowland

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: bc05f26fed912144c57cb442472e777f0f203b51
Build: success
Install hpccsystems-platform-community_7.0.0-rc2.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |806 |806 |0 |
|test (thor) |730 |730 |0 |
|test (roxie) |879 |879 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   13 sec (00:00:13) |  186 sec (00:03:06) |   95 sec (00:01:35) |   20 sec (00:00:20) |   42 sec (00:00:42) | 1359 sec (00:22:39) |   18 sec (00:00:18) | 1733 sec (00:28:53) |

richardkchapman commented:
Needs rebasing/squashing

richardkchapman committed:
HPCC-20332 Initial Dali XSD

Convert Dali to new format
Change XSDs to use new default custom namespace attributes
Add hpcc:hiddenIf custom namespace attribute

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

kenrowland commented:
@richardkchapman Please merge

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 8763dc02e47bb477f8ece503bedea32063385b32
Build: success
Install hpccsystems-platform-community_7.0.0-rc2.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |806 |806 |0 |
|test (thor) |730 |730 |0 |
|test (roxie) |879 |879 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   25 sec (00:00:25) |  332 sec (00:05:32) |  147 sec (00:02:27) |   19 sec (00:00:19) |   41 sec (00:00:41) | 1540 sec (00:25:40) |   20 sec (00:00:20) | 2124 sec (00:35:24) |
